File: src/converters/VisaConverter.ts
import { FileConverter, TransactionRecord } from './FileConverter';

export class VisaConverter implements FileConverter {
  private readonly VISA_LINE_REGEX = /^(\d{8}),(\d+\.\d{2}),(.{25}),(\d{16}),(\w{12})$/;

  validate(fileContent: string): boolean {
    const lines = fileContent.split('\n');
    return lines.every(line => this.VISA_LINE_REGEX.test(line.trim()));
  }

  parse(fileContent: string): TransactionRecord[] {
    if (!this.validate(fileContent)) {
      throw new Error('Invalid VISA file format');
    }

    const lines = fileContent.split('\n');
    return lines.map(line => {
      const [date, amount, merchant, card, txId] = line.trim().split(',');
      
      return {
        transactionDate: this.parseDate(date),
        amount: parseFloat(amount),
        merchantName: merchant.trim(),
        cardNumber: card,
        transactionId: txId
      };
    });
  }

  private parseDate(dateStr: string): Date {
    // Assuming date format is YYYYMMDD
    const year = parseInt(dateStr.substring(0, 4));
    const month = parseInt(dateStr.substring(4, 6)) - 1;
    const day = parseInt(dateStr.substring(6, 8));
    return new Date(year, month, day);
  }
} 

File: converters/visa_converter.py
from datetime import datetime
import re

class VisaConverter:
    def __init__(self):
        # Updated pattern to be more flexible with whitespace
        self.line_pattern = re.compile(r'^(\d{8}),(\d+\.\d{2}),(.+?),(\d{16}),(\w{12})$')

    def parse(self, file_content: str) -> list:
        transactions = []
        for line in file_content.strip().split('\n'):
            match = self.line_pattern.match(line.strip())
            if not match:
                raise ValueError(f"Invalid VISA format: {line}")
            
            date_str, amount, merchant, card, tx_id = match.groups()
            
            # Parse date (YYYYMMDD format)
            date = datetime.strptime(date_str, '%Y%m%d')
            
            transactions.append({
                'transaction_date': date,
                'amount': float(amount),
                'merchant_name': merchant.strip(),
                'card_number': card,
                'transaction_id': tx_id
            })
        
        return transactions 